executor/read/pattern_executor.rs

```diff
@@ -0,0 +1,188 @@
+                        let row = controller.prepare_and_map_input(&unmapped_input);
+                        for (branch_index, pattern) in branches.iter_mut().enumerate() {
+                            pattern.prepare(row.clone().into_owned());
+                            control_stack
+                                .push(ControlInstruction::ExecuteSubQuery(index.clone(), BranchIndex(branch_index)));
+                        }
+                    }
+                }
+                ControlInstruction::ExecuteSubQuery(index, branch_index) => {
+                    // TODO: This bit desperately needs a cleanup.
+                    let (branches, controller) = &mut executors[index.0].unwrap_nested_pattern_branch().to_parts_mut();
+                    let branch = &mut branches[branch_index.0];
+                    let unmapped = branch.pattern_executor.batch_continue(context, interrupt)?;
+                    let found = match controller.map_output(branch.input.as_ref().unwrap(), unmapped) {
+                        SubQueryResult::Retry(found) => {
+                            control_stack.push(ControlInstruction::ExecuteSubQuery(index.clone(), branch_index));
+                            found
+                        }
+                        SubQueryResult::Done(found) => {
+                            branch.pattern_executor.reset();
+                            found
+                        }
+                    };
+                    if let Some(batch) = found {
+                        self.prepare_next_instruction_and_push_to_stack(context, index.next(), batch)?;
+                    }
+                }
+                ControlInstruction::CollectingStage(index) => {
+                    let (pattern, mut collector) = executors[index.0].unwrap_collecting_stage().to_parts_mut();
+                    match pattern.compute_next_batch(context, interrupt)? {
+                        Some(batch) => {
+                            collector.accept(context, batch);
+                            self.control_stack.push(ControlInstruction::CollectingStage(index))
+                        }
+                        None => {
+                            let iterator = collector.into_iterator();
+                            self.control_stack.push(ControlInstruction::StreamCollected(index.clone(), iterator))
+                        }
+                    }
+                }
+                ControlInstruction::StreamCollected(index, mut collected_iterator) => {
+                    if let Some(batch) = collected_iterator.batch_continue()? {
+                        self.control_stack.push(ControlInstruction::StreamCollected(index.clone(), collected_iterator));
+                        self.prepare_next_instruction_and_push_to_stack(context, index.next(), batch)?;
+                    }
+                }
+                ControlInstruction::Yield(batch) => {
+                    return Ok(Some(batch));
+                }
+            }
```

Comment:
I'm tempted to refactor this into a `batch_continue` which produces the `Option<Batch>` + a `transition` function based on the current `ControlInstruction`